dot/sync/chain_processor.go

```diff
@@ -0,0 +1,278 @@
+		if errors.Is(err, blocktree.ErrBlockExists) {
+			return nil
+		} else if err != nil {
+			logger.Warn("failed to add block to blocktree", "hash", bd.Hash, "error", err)
+			return err
+		}
+
+		if bd.Justification != nil {
+			logger.Debug("handling Justification...", "number", block.Header.Number, "hash", bd.Hash)
+			s.handleJustification(&block.Header, *bd.Justification)
+		}
+
+		// TODO: this is probably unnecessary, since the state is already in the database
+		// however, this case shouldn't be hit often, since it's only hit if the node state
+		// is rewinded or if the node shuts down unexpectedly
+		state, err := s.storageState.TrieState(&block.Header.StateRoot)
+		if err != nil {
+			logger.Warn("failed to load state for block", "block", block.Header.Hash(), "error", err)
+			return err
+		}
+
+		if err := s.blockImportHandler.HandleBlockImport(block, state); err != nil {
+			logger.Warn("failed to handle block import", "error", err)
+		}
+
+		return nil
+	}
+
+	if bd.Header != nil && bd.Body != nil {
+		if err = s.handleHeader(bd.Header); err != nil {
+			return err
+		}
+
+		if err = s.handleBody(bd.Body); err != nil {
+			return err
+		}
+
+		block := &types.Block{
+			Header: *bd.Header,
+			Body:   *bd.Body,
+		}
+
+		logger.Debug("processing block", "hash", bd.Hash)
+
+		if err = s.handleBlock(block); err != nil {
+			logger.Error("failed to handle block", "number", block.Header.Number, "error", err)
+			return err
+		}
+
+		logger.Debug("block processed", "hash", bd.Hash)
+	}
+
+	if bd.Justification != nil && bd.Header != nil {
+		logger.Debug("handling Justification...", "number", bd.Number(), "hash", bd.Hash)
+		s.handleJustification(bd.Header, *bd.Justification)
+	}
+
+	return nil
+}
+
+// handleHeader handles headers included in BlockResponses
+func (s *chainProcessor) handleHeader(header *types.Header) error {
+	err := s.babeVerifier.VerifyBlock(header)
+	if err != nil {
+		return fmt.Errorf("%w: %s", ErrInvalidBlock, err.Error())
+	}
+
+	return nil
+}
+
+// handleHeader handles block bodies included in BlockResponses
+func (s *chainProcessor) handleBody(body *types.Body) error {
+	exts, err := body.AsExtrinsics()
+	if err != nil {
+		return fmt.Errorf("cannot parse body as extrinsics: %w", err)
+	}
+
+	for _, ext := range exts {
+		s.transactionState.RemoveExtrinsic(ext)
+	}
+
+	return err
+}
+
+// handleHeader handles blocks (header+body) included in BlockResponses
+func (s *chainProcessor) handleBlock(block *types.Block) error {
+	if block == nil || block.Body == nil {
+		return errors.New("block, header, or body is nil")
+	}
+
+	parent, err := s.blockState.GetHeader(block.Header.ParentHash)
+	if err != nil {
+		return fmt.Errorf("failed to get parent hash: %w", err)
+	}
+
+	s.storageState.Lock()
+	defer s.storageState.Unlock()
+
+	ts, err := s.storageState.TrieState(&parent.StateRoot)
+	if err != nil {
+		return err
+	}
+
+	root := ts.MustRoot()
+	if !bytes.Equal(parent.StateRoot[:], root[:]) {
```

Comment:
why we compare the root (generated from `parent.StateRoot`) with the `parent.StateRoot`?

Reply:
the `root` is loaded from the `StorageState` using the parent block's state root, we want the trie to have the correct hash (ie have all the correct data) before proceeding. if there's ever a case where they don't match, we have a major bug in the code and need to panic